### PR TITLE
@kadena/transfer

### PR DESCRIPTION
**Pull request by MRVDH**
We are merging the kadena transfer app into the main branch. Note: this app is not finished yet.

### CODE REVIEW

File: common/config/rush/pnpm-lock.yaml [hunk 1]
```diff
@@ -4857,12 +5255,24 @@ packages:
     engines: {node: ^10.13.0 || ^12.13.0 || ^14.15.0 || >=15.0.0}
     dependencies:
       '@jest/types': 27.5.1
-      '@types/node': 16.18.7
+      '@types/node': 18.11.13
       chalk: 4.1.2
```
Comment (alber70g) [inline common/config/rush/pnpm-lock.yaml:5296]: We support node 16 and 18, does this potentially break on node 16

File: common/config/rush/pnpm-lock.yaml [hunk 2]
```diff
@@ -5121,12 +5607,12 @@ packages:
...
-  /@jest/schemas/29.0.0:
-    resolution: {integrity: sha512-3Ab5HgYIIAnS0HjqJHQYZS+zXc4tUmTmBH3z83ajI6afXp8X3ZtdLX+nXx+I7LNkJD7uN9LAVhgnjDgZa2z0kA==}
-    engines: {node: ^14.15.0 || ^16.10.0 || >=18.0.0}
+  /@jest/schemas/28.1.3:
+    resolution: {integrity: sha512-/l/VWsdt/aBXgjshLWOFyFt3IVdYypu5y2Wn2rOO1un6nkqIn8SLXzgIMYXFyYsRWDyF5EthmKJMIdJvk08grg==}
+    engines: {node: ^12.13.0 || ^14.15.0 || ^16.10.0 || >=17.0.0}
     dependencies:
       '@sinclair/typebox': 0.24.51
-    dev: true
+    dev: false
```
Comment (alber70g) [inline common/config/rush/pnpm-lock.yaml:5653]: Shouldn't this be `dev: true`?

File: packages/apps/transfer/.gitignore
Comment (alber70g) [inline packages/apps/transfer/.gitignore:1]: Can we dedupe everything that's already ignored in the root `.gitignore`? And the things that can be ignored everywhere can be added to the root `.gitignore`, like `.idea/*`

File: packages/apps/transfer/README.md
Comment (alber70g) [inline packages/apps/transfer/README.md:1]: Can we add a list of features that the transfer app offers?

And maybe create it's own section for developers **Contributing** or something.
And remove the redundant stuff like docs on api routes, etc

File: packages/apps/transfer/package.json [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    "build": "next build",
+    "dev": "next dev",
+    "eject": "react-scripts eject",
+    "generate:icons": "npx @svgr/cli --out-dir src/resources/svg/generated -- src/resources/svg",
+    "postinstall": "rushx generate:icons",
+    "lint": "next lint",
+    "lint-staged": "lint-staged",
+    "lint:fix": "eslint --fix .",
+    "start": "next start",
+    "test": ""
```
Comment (alber70g) [inline packages/apps/transfer/package.json:17]: I'd like to have some tests (maybe to add later on). 

Doesn't need to be high coverage, but to test the base features with happy flows. This gives us more certainty when releasing new versions of underlying libs as well

File: packages/apps/transfer/src/components/Global/Ace/index.tsx [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+import 'ace-builds/src-noconflict/mode-clojure';
+import 'ace-builds/src-noconflict/mode-lisp';
+import 'ace-builds/src-noconflict/theme-monokai';
+import 'ace-builds/src-noconflict/theme-one_dark';
+
+export interface AceEditorProps {
+  code?: string;
+}
+
+const AceViewerComponent: FC<AceEditorProps> = ({ code }) => {
```
Comment (alber70g) [inline packages/apps/transfer/src/components/Global/Ace/index.tsx:14]: Question: I haven't started the app so maybe this is irrelevant. Do we expose the option to enable vim-mode, and maybe other useful QoL features?

File: packages/apps/transfer/src/pages/_app.tsx [hunk 5]
```diff
@@ -0,0 +1,14 @@
+import '@/resources/styles/globals.css';
+
+import { Layout } from '@/components/Common';
+import type { AppProps } from 'next/app';
+import React, { FC } from 'react';
+
+const App: FC<AppProps> = ({ Component, pageProps }: AppProps) => (
+  <Layout>
+    {/* @ts-ignore */}
```
Comment (alber70g) [inline packages/apps/transfer/src/pages/_app.tsx:9]: Is this required?

File: packages/apps/transfer/src/pages/module-explorer/index.tsx [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+              value={moduleChain}
+            >
+              {renderChainOptions()}
+            </Select>
+            <TextField
+              label={t('Module Name')}
+              inputProps={{
+                placeholder: t('Enter desired module name'),
+                // @ts-ignore
+                onChange: (e) => setModuleName(e?.target?.value),
```
Comment (alber70g) [inline packages/apps/transfer/src/pages/module-explorer/index.tsx:88]: Why is this necessary?

File: packages/apps/transfer/src/services/accounts/get-balance.ts [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+  const arrPromises = [];
+
+  const pactCommand = new PactCommand();
+
+  pactCommand.code = createExp(`${token}.details "${accountName}"`);
+
+  try {
+    for (let i = 0; i < 20; i++) {
+      const NETWORK_ID = server.includes('testnet') ? 'testnet04' : 'mainnet01';
+      const API_HOST = `https://${server}/chainweb/0.0/${NETWORK_ID}/chain/${i}/pact`;
```
Comment (alber70g) [inline packages/apps/transfer/src/services/accounts/get-balance.ts:33]: Idem above. Use these from envvar?

File: packages/apps/transfer/src/services/modules/list-module.ts [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+import { createExp } from '@kadena/pactjs';
+
+//List modules PACT code example
+
+const NETWORK_ID: any = 'testnet04';
+const chainId: string = '1';
+const gasLimit: number = 6000;
+const gasPrice: number = 0.001;
+const ttl: number = 600;
+export const API_HOST: string = `https://api.testnet.chainweb.com/chainweb/0.0/${NETWORK_ID}/chain/${chainId}/pact`;
```
Comment (alber70g) [inline packages/apps/transfer/src/services/modules/list-module.ts:11]: Idem. Env vars?

File: packages/apps/transfer/src/services/transfer/coin-transfer.ts [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+
+  const toAccountGuard = await getPublicKeys(networkId, fromChainId, toAccount);
+  if (toAccountGuard === undefined) {
+    throw new Error('Receiver account does not exist');
+  }
+
+  const pactCommand = new PactCommand();
+  pactCommand.code = `(coin.transfer-create "${fromAccount}" "${toAccount}" (read-keyset "ks") ${new PactNumber(
+    amount,
+  ).toDecimal()})`;
```
Comment (alber70g) [inline packages/apps/transfer/src/services/transfer/coin-transfer.ts:94]: Not sure if we should at this moment, but it could be nice to use `Pact.modules.coin.transfer` here with the generated types?

(Applies to other functions as well)

File: packages/apps/transfer/tsconfig.json [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+    "isolatedModules": true,
+    "declaration": false, // fixes pnpm issue: https://github.com/microsoft/TypeScript/issues/42873
+    "declarationMap": false,
+    "jsx": "preserve",
+    "paths": {
+      "@/*": ["./src/*"]
+    }
+  },
+  "exclude": ["node_modules"]
+}
```
Comment (alber70g) [inline packages/apps/transfer/tsconfig.json:24]: Is this the minimum that we need when based on the `@kadena-dev/heft-rig/profiles/default/tsconfig-base.json`
If it is, and it's required for all next apps that we have, we could make a common `tsconfig-base.json` specifically for `next`. (ie: `@kadena-dev/heft-rig/profiles/nextjs/tsconfig-base.json`

**Review by alber70g — CHANGES_REQUESTED** [review]

File: common/config/rush/pnpm-lock.yaml (hunk 1 above)
Comment (MRVDH) [reply common/config/rush/pnpm-lock.yaml:5296]: It's generally more the other way around (due to backwards compatibility), if we put 16 here but also support 18, it could throw errors since it's not up to date with the newest types.

File: packages/apps/transfer/.gitignore
Comment (rezaerami) [reply packages/apps/transfer/.gitignore:1]: done

File: packages/apps/transfer/README.md
Comment (rezaerami) [reply packages/apps/transfer/README.md:1]: done

File: packages/apps/transfer/tsconfig.json (hunk 10 above)
Comment (rezaerami) [reply packages/apps/transfer/tsconfig.json:24]: new ts config for next is a bit different, I removed duplicates, but the remaining ts config is suitable for new next version

File: rush.json [hunk 11]
```diff
@@ -424,6 +424,14 @@
       "shouldPublish": false
     },
 
+    {
+      "packageName": "@kadena/transfer",
+      "projectFolder": "packages/apps/transfer",
+      "reviewCategory": "prototypes",
+      "skipRushCheck": true,
```
[the remark this answers is not in the document]
Comment (rezaerami) [reply rush.json:431]: due to issues that we had we had to skip it, for now I think it works if we bring the rush check back again
so I will turn it on again

File: packages/apps/transfer/src/services/modules/list-module.ts (hunk 8 above)
Comment (rezaerami) [reply packages/apps/transfer/src/services/modules/list-module.ts:11]: i will implement a whole config service for this one soon

File: packages/apps/transfer/src/services/transfer/coin-transfer.ts (hunk 9 above)
Comment (MRVDH) [reply packages/apps/transfer/src/services/transfer/coin-transfer.ts:94]: Most of this code will be removed this week since we are not doing coin transfers anymore. If we do still have some of this code by then we will check it out 👍

File: packages/apps/transfer/src/pages/coin-transfer/index.tsx [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+import { convertIntToChainId } from '@/services/utils/utils';
+import useTranslation from 'next-translate/useTranslation';
+import React, { FC, useState } from 'react';
+
+const CoinTransfer: FC = () => {
+  const { t } = useTranslation('common');
+  const NETWORK_ID = 'testnet04';
+  const chainId = '1';
+  const API_HOST = `https://api.testnet.chainweb.com/chainweb/0.0/${NETWORK_ID}/chain/${chainId}/pact`;
+  const numberOfChains = 20;
```
[the remark this answers is not in the document]
Comment (MRVDH) [reply packages/apps/transfer/src/pages/coin-transfer/index.tsx]: Relevant code removed

File: packages/apps/transfer/src/services/accounts/get-balance.ts [hunk 13]
```diff
@@ -0,0 +1,62 @@
+import { PactCommand } from '@kadena/client';
+import { createExp } from '@kadena/pactjs';
+import { type ChainId } from '@kadena/types';
+
+const gasLimit: number = 6000;
+const gasPrice: number = 0.001;
+const ttl: number = 600;
+
+// eslint-disable-next-line @typescript-eslint/naming-convention
+export interface ChainResult {
```
[the remark this answers is not in the document]
Comment (MRVDH) [reply packages/apps/transfer/src/services/accounts/get-balance.ts:10]: Relevant code removed

File: packages/apps/transfer/src/services/accounts/get-public-keys.ts [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+import { PactCommand } from '@kadena/client';
+import { createExp } from '@kadena/pactjs';
+import { type ChainId } from '@kadena/types';
+
+import { generateApiHost } from '../utils/utils';
+const gasLimit: number = 6000;
+const gasPrice: number = 0.001;
+const ttl: number = 600;
+// eslint-disable-next-line @typescript-eslint/naming-convention
+export interface AccountGuard {
```
[the remark this answers is not in the document]
Comment (MRVDH) [reply packages/apps/transfer/src/services/accounts/get-public-keys.ts]: Outdated code, removed

File: packages/apps/transfer/src/pages/_app.tsx (hunk 5 above)
Comment (MRVDH) [reply packages/apps/transfer/src/pages/_app.tsx]: Removed

File: packages/apps/transfer/src/services/accounts/get-balance.ts (hunk 7 above)
Comment (MRVDH) [reply packages/apps/transfer/src/services/accounts/get-balance.ts]: Outdated since this code does not exist anymore

File: packages/apps/transfer/src/pages/module-explorer/index.tsx (hunk 6 above)
Comment (MRVDH) [reply packages/apps/transfer/src/pages/module-explorer/index.tsx]: Fixed

File: packages/apps/transfer/src/components/Global/Ace/index.tsx (hunk 4 above)
Comment (MRVDH) [reply packages/apps/transfer/src/components/Global/Ace/index.tsx:14]: We don't need it to be an editor, just a code viewer, therefore we kept it like this

File: packages/apps/transfer/README.md [hunk 15]
```diff
@@ -1,38 +1,99 @@
...
+- [License](#license)
+
+## Installation
+
+To install and run the Kadena Transfer App locally, follow these steps:
+
+1. Clone the repository:
+
+```bash
+git clone https://github.com/kadena-community/kadena.js.git
```
Comment (alber70g) [inline packages/apps/transfer/README.md:19]: Nitpick: why not `ssh` clone?

File: packages/apps/transfer/README.md [hunk 16]
```diff
@@ -1,38 +1,99 @@
...
+2. Navigate to the `kadena-transfer/packages/apps/transfer` directory:
+
+```bash
+cd kadena.js/kadena-transfer/packages/apps/transfer
+```
+
+3. Install the dependencies:
+
+```bash
+rush install
```
Comment (alber70g) [inline packages/apps/transfer/README.md:31]: Should we have a step to install rush globally? `npm install -g @microsoft/rush`

Nitpick: I like the step approach, I don't like that I cant copy paste this into the shell directly. I'd opt for

```bash
# install rush
npm install --global @microsoft/rush
# clone repo
git clone git@github.com:kadena-community/kadena.js.git
cd kadena.js/kadena-transfer/packages/apps/transfer
# install dependencies
rush install
# build app with dependencies
rush install --to-except @kadena/transfer
```

Start the app
```bash
rushx dev
```

File: packages/apps/transfer/README.md [hunk 17]
```diff
@@ -1,38 +1,99 @@
...
+
+5. Start the app:
 
 ```bash
-npm run dev
-# or
-yarn dev
-# or
-pnpm dev
+rush dev
```
Comment (alber70g) [inline packages/apps/transfer/README.md:43]: `rushx dev`?

File: packages/apps/transfer/README.md [hunk 18]
```diff
@@ -1,38 +1,99 @@
...
 ```
 
-Open [http://localhost:3000](http://localhost:3000) with your browser to see the result.
 
-You can start editing the page by modifying `pages/index.tsx`. The page auto-updates as you edit the file.
+The Kadena Transfer App will be accessible at `http://localhost:3000` in your web browser.
```
Comment (alber70g) [inline packages/apps/transfer/README.md:47]:
```suggestion
The Kadena Transfer App will be accessible at [http://localhost:3000](http://localhost:3000) in your web browser.
```

File: packages/apps/transfer/README.md
Comment (alber70g) [inline packages/apps/transfer/README.md:1]: Is the pre-commit staged executed? We have a max line lenght and `prosewrap` in prettier

File: packages/apps/transfer/README.md [hunk 19]
```diff
@@ -1,38 +1,99 @@
...
+
+Together, these features empower users to interact with the blockchain network, manage their identities, monitor account balances, transfer digital assets, and explore the blockchain's data.
+
+## Contributing
+
+Contributions to the Kadena Transfer App are welcome! If you find a bug, have a feature request, or want to contribute code, please follow these steps:
+
+1. Fork the repository.
+
+2. Create a new branch for your feature or bug fix:
```
Comment (alber70g) [inline packages/apps/transfer/README.md:74]: We could add to create an issue first:

```suggestion
1. [Report the issue or feature request](https://github.com/kadena-community/kadena.js/issues/new/choose)
2. Fork the repository.
3. Create a new branch for your feature or bug fix:
```

File: packages/apps/transfer/tsconfig.json [hunk 20]
```diff
@@ -6,15 +6,12 @@
     "lib": ["dom", "dom.iterable", "esnext"],
     "allowJs": true,
```
Comment (alber70g) [inline packages/apps/transfer/tsconfig.json:7]: Where are we using `js`?

**Review by alber70g — APPROVED** [review]

File: packages/apps/transfer/src/services/key-pairs/key-pairs.ts [hunk 21]
```diff
@@ -0,0 +1,12 @@
...
+import { downloadFileToBrowser } from '../utils/file-download';
+
+export function downloadKeyPairToBrowser(): void {
+  const keyPair = genKeyPair();
+
+  downloadFileToBrowser(
+    'key-pair.txt',
+    `public: ${keyPair.publicKey}\r\nprivate: ${keyPair.secretKey}`,
+  );
+}
```
Comment (alber70g) [inline packages/apps/transfer/src/services/key-pairs/key-pairs.ts:12]: Idea: could we change the textfile to be `kadena-<pubkey>.txt`? That way we don't inadvertently overwrite an existing key

File: .gitignore [hunk 22]
```diff
@@ -85,6 +85,7 @@ common/autoinstallers/*/.npmrc
 **/.rush/temp/
 **/.rush
 packages/*/*/temp
+rush-logs
```
[the remark this answers is not in the document]
Comment (MRVDH) [reply .gitignore:88]: Fixed 👍

File: packages/apps/transfer/README.md (hunk 17 above)
Comment (MRVDH) [reply packages/apps/transfer/README.md:43]: Changed 👍

File: packages/apps/transfer/README.md (hunk 19 above)
Comment (MRVDH) [reply packages/apps/transfer/README.md:74]: Added 👍

File: packages/apps/transfer/README.md (hunk 18 above)
Comment (MRVDH) [reply packages/apps/transfer/README.md:47]: Fixed 👍

File: packages/apps/transfer/README.md [hunk 23]
```diff
@@ -1,38 +1,99 @@
...
+3. Install the dependencies:
+
+```bash
+rush install
+```
+
+4. Build the app and its dependencies:
+
+```bash
+rush build
```
[the remark this answers is not in the document]
Comment (MRVDH) [reply packages/apps/transfer/README.md:37]: Fixed 👍

File: packages/apps/transfer/README.md (hunk 15 above)
Comment (MRVDH) [reply packages/apps/transfer/README.md:19]: Since the user can pick any type they want (https/ssh/ghcli) I changed the link to the repo page. 👍

File: packages/apps/transfer/README.md
Comment (MRVDH) [reply packages/apps/transfer/README.md:1]: Yes 👍

File: packages/apps/transfer/tsconfig.json (hunk 20 above)
Comment (MRVDH) [reply packages/apps/transfer/tsconfig.json:7]: Nope, can be removed indeed. (we used to, due to an outdated version of react and some autogenerated files)

**Review by webpro — COMMENTED** [review]
Looking at the [CONTRIBUTING.md](https://github.com/kadena-community/kadena.js/blob/master/CONTRIBUTING.md) we already have, what was your idea about the relationship with the README.md this package?

Some ideas I have:

- Packages are stand-alone and are managed separately by their code owners.
- We collaborate and link to the "root" one as much as possible (DRY).
- We collaborate and use #384 to "install" directives/comment markers and generate generic instructions everywhere.

Looking at the number of repositories and packages we have I'd say the first one isn't the best option.

WDYT?

**Review by KristinaSpasevska — APPROVED** [review]

**Review by alber70g — APPROVED** [review]
